pkg/api/dashboard.go

```diff
@@ -306,6 +308,26 @@ func (hs *HTTPServer) PostDashboard(c *models.ReqContext) response.Response {
 	if err := web.Bind(c.Req, &cmd); err != nil {
 		return response.Error(http.StatusBadRequest, "bad request data", err)
 	}
+
+	// Ideally, coremodel validation calls would be integrated into the web
+	// framework. But this does the job for now.
+	if cm, has := hs.CoremodelRegistry.Get("dashboard"); has {
+		schv, err := cmd.Dashboard.Get("schemaVersion").Int()
+
+		// Only try to validate if the schemaVersion is at least the handoff version
+		// (the minimum schemaVersion against which the dashboard schema is known to
+		// work), or if schemaVersion is absent (which will happen once the Thema
+		// schema becomes canonical).
+		if err != nil || schv >= dashboard.HandoffSchemaVersion {
+			// Can't fail, web.Bind() already ensured it's valid JSON
+			b, _ := cmd.Dashboard.Bytes()
+			v, _ := cuectx.JSONtoCUE("dashboard.json", b)
+			if _, err := cm.CurrentSchema().Validate(v); err != nil {
+				return response.Error(http.StatusBadRequest, "invalid dashboard json", err)
+			}
+		}
```

Comment:
Yeah, this PR has definitely led me to wanting some more abstractions here. I've definitely thought of the "can this entity in particular be validated" case - i wrote the [same comparison last week](https://github.com/grafana/grafana/blob/484bd0b55748ac33b0975eba55e201a1394f0d1d/devenv/dev-dashboards/dashboards_test.go#L71). For that, i was pondering adding a `CanValidate([]byte)` method to `coremodel.Interface()`. I'm not sure we'd ever return anything but `true` for it for any coremodel other than dashboards, but that's OK.

But, combining two of your ideas: the notion of some outer, registry-wrapping `Validate()` call (or other `coremodel.Interface` method), plus tying feature flags to those particular methods? That's brilliant. I _really_ like that.

That said, i think there's a lot to think about there. I'd prefer we get in this initial, flawed implementation so that we have a baseline, then can iteratively abstract.